### ISSUE TITLE
Comparing Time and integer does not result in a type error

### ISSUE THREAD

**Issue by alanmcgovern** [sorbet/sorbet#3278]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Arequire%20'time'%0A%0Adef%20main%0A%20%20value%20%3D%20%7Bval%3A123%7D%5B'val'%5D.to_i%0A%20%20(value%20-%2030)%20%3C%20Time.now%0Aend)

```ruby
# typed: true
require 'time'

def main
  value = {val:123}[:val].to_i
  (value - 30) < Time.now
end
```

#### Observed output

No error is displayed

#### Expected behavior
`value` should not be `T.untyped` due to the `to_i` invocation, and this should show a type error due to incorrectly comparing an `Integer` and a `Time`

- - -

**Comment by elliottt** [comment]
This looks like another bug with shape-types.

**Comment by karmakaze** [comment]
The problem is closer to applying `.to_i` to a `nil` value.
```
  value = nil.to_i
  T.reveal_type(value) # T.untyped

  T.reveal_type({val:123}[:val]) # T.untyped
```
The last one is a known problem with literal Shapes where even though the type could be known to be non-nil, the hash[] lookup is considered to return nilable values. That issue could be fixed by #3823

**Comment by jez** [comment]
I've put up a PR for the `NilClass#to_i` sig: #6183 

The other bug is simply that hash literals are treated as shapes, but shape support is very bad. That is tracked by #713.
